### PR TITLE
`generate` v0.0.1

### LINKED ISSUE #1: Add `deno generate` proof-of-concept
Dengen is a proof-of-concept command-line interface following the design of my formal proposal for a `deno generate` subcommand, https://github.com/denoland/deno/issues/19176.

### PR DESCRIPTION
**Pull request by EthanThatOneKid**
This is a large PR intended to resolve <https://github.com/EthanThatOneKid/dengen/issues/1>.

It consists of a module (`mod.ts`) and a CLI tool (`main.ts`).

### CODE REVIEW

File: examples/with_generate_docs.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+//deno:generate -command gen_docs deno run -A generate_docs.ts
```
Comment (NfNitLoop) [inline examples/with_generate_docs.ts]: typically named parameters use `--` as a prefix, to disambiguate them from parameters that are a combination of one-letter parameters (ex: `tar -cvf`)

File: parse/parse.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+ * because to run arbitrary code generation commands, we only
+ * need to parse the file to determine the comments. With each
+ * comment, we return the line and character of the comment so that
+ * we can tell the user where errors are.
+ *
+ * Implements dynamic programming to get the line and character of the comment.
+ */
+export async function parseComments(reader: Reader): Promise<ParsedComment[]> {
+  const comments: ParsedComment[] = [];
+  let currentLine = BigInt(0);
```
Comment (NfNitLoop) [inline parse/parse.ts]: BigInt is probably unnecessary here.  You're unlikely to ever parse a file with more than 9007199254740991 lines. And if you care about that case, you might just want to skip files that are over that many bytes long. 😊

**Review by NfNitLoop — COMMENTED** [review]
Didn't have time to review every line, but the shape of things is good.  Good docs, especially on your exported interfaces. 👍 

My main questions are about the design:

* Why do I want to put `//deno:generate <command>` in a TypeScript file instead of just in `deno task`? 
* If I have two files that depend on some generated code, which file do I put the annotation in?

**Comment by EthanThatOneKid** [comment]
Thanks for taking the time to review, @NfNitLoop! I appreciate your feedback. Allow me to address the main questions.

> Why do I want to put `//deno:generate <command>` in a TypeScript file instead of just in `deno task`?

Using `//deno:generate <command>` in a TypeScript file instead of solely relying on `deno task` offers several advantages. Firstly, incorporating `//deno:generate` directly into the source code allows for better integration between the generators and the codebase. This approach enables developers to easily understand and manage the generators as an integral part of the project.

Additionally, by utilizing `//deno:generate`, we can conveniently run multiple generators that are closely tied to the specific TypeScript files. While it's technically feasible to write a `deno task` that mimics the functionality of a series of `//deno:generate` commands, this approach may not scale efficiently when the project relies on numerous generators. By placing the `//deno:generate` directives within the relevant TypeScript files, we achieve a more scalable and flexible solution for managing generator-related tasks.

> If two files rely on some generated code, it is recommended to make a shared module from which the two files can import the generated code.

In situations where multiple files depend on the same generated code, it's highly recommended to create a shared module. This shared module serves as a central location from which both files can import the generated code. By doing so, we promote better code organization and encourage code reuse.

The shared module approach also simplifies future modifications and updates. If the generated code needs to be modified or enhanced, we can make the changes in a single place—the shared module—and have the updates reflect in both files that depend on it. This not only reduces code duplication but also improves maintainability by ensuring consistency throughout the project.

I hope this clarifies the benefits of using `//deno:generate` in TypeScript files and the advantages of employing a shared module when dealing with generated code dependencies. Please let me know if you have any further questions or concerns.

File: examples/with_generate_docs.ts (hunk 1 above)
Comment (EthanThatOneKid) [reply examples/with_generate_docs.ts]: > typically named parameters use `--` as a prefix, to disambiguate them from parameters that are a combination of one-letter parameters (ex: `tar -cvf`)

The `-command` alias in `//go:generate -command` is a way to create an alias for a command that you want to run later. It allows you to define the command once and then use the alias multiple times in the file it was defined in, appending different arguments to the alias command when you need to run it. The `//deno:generate -command` directive is inspired by  `//go:generate -command`.

File: parse/parse.ts (hunk 2 above)
Comment (EthanThatOneKid) [reply parse/parse.ts]: > BigInt is probably unnecessary here. You're unlikely to ever parse a file with more than 9007199254740991 lines. And if you care about that case, you might just want to skip files that are over that many bytes long. 😊

I agree that using BigInt is unnecessary in this context. Considering the extremely rare case of parsing a file with more than 9007199254740991 lines, it's reasonable to skip files that exceed that length. I have gone ahead and committed the suggested changes <https://github.com/EthanThatOneKid/deno_generate/pull/1/commits/dffc7a0213afc3db91007f0b9693df72f93a0ec9>.

**Pull request merged**